lib/core/src/recover/model.rs

```diff
@@ -0,0 +1,684 @@
+    ///
+    /// -1 means unconfirmed with unconfirmed parents
+    ///  0 means unconfirmed with confirmed parents
+    pub height: i32,
+}
+impl HistoryTxId {
+    pub(crate) fn confirmed(&self) -> bool {
+        self.height > 0
+    }
+}
+impl From<History> for HistoryTxId {
+    fn from(value: History) -> Self {
+        Self::from(&value)
+    }
+}
+impl From<&History> for HistoryTxId {
+    fn from(value: &History) -> Self {
+        Self {
+            txid: value.txid,
+            height: value.height,
+        }
+    }
+}
+
+/// A map of all our known LWK onchain txs, indexed by tx ID. Essentially our own cache of the LWK txs.
+pub(crate) struct TxMap {
+    pub(crate) outgoing_tx_map: HashMap<Txid, WalletTx>,
+    pub(crate) incoming_tx_map: HashMap<Txid, WalletTx>,
+}
+impl TxMap {
+    pub(crate) fn from_raw_tx_map(raw_tx_map: HashMap<Txid, WalletTx>) -> Self {
+        let (outgoing_tx_map, incoming_tx_map): (HashMap<Txid, WalletTx>, HashMap<Txid, WalletTx>) =
+            raw_tx_map
+                .into_iter()
+                .partition(|(_, tx)| tx.balance.values().sum::<i64>() < 0);
+
+        Self {
+            outgoing_tx_map,
+            incoming_tx_map,
+        }
+    }
+}
+
+pub(crate) trait PartialSwapState {
+    /// Determine partial swap state, based on recovered chain data.
+    ///
+    /// This is a partial state, which means it may be incomplete because it's based on partial
+    /// information. Some swap states cannot be determined based only on chain data.
+    /// In these cases we do not assume any swap state.
+    fn derive_partial_state(&self) -> Option<PaymentState>;
+}
+
+pub(crate) struct RecoveredOnchainDataSend {
+    pub(crate) lockup_tx_id: Option<HistoryTxId>,
+    pub(crate) claim_tx_id: Option<HistoryTxId>,
+    pub(crate) refund_tx_id: Option<HistoryTxId>,
+}
+impl PartialSwapState for RecoveredOnchainDataSend {
+    fn derive_partial_state(&self) -> Option<PaymentState> {
+        match &self.lockup_tx_id {
+            Some(_) => match &self.claim_tx_id {
+                Some(_) => Some(PaymentState::Complete),
+                None => match &self.refund_tx_id {
+                    Some(refund_tx_id) => match refund_tx_id.confirmed() {
+                        true => Some(PaymentState::Failed),
+                        false => Some(PaymentState::RefundPending),
+                    },
+                    None => Some(PaymentState::Pending),
+                },
+            },
+            // We have no onchain data to support deriving the state as the swap could
+            // potentially be Created, TimedOut or Failed after expiry. In this case we return None.
+            None => None,
+        }
+    }
+}
+
+pub(crate) struct RecoveredOnchainDataReceive {
+    pub(crate) lockup_tx_id: Option<HistoryTxId>,
+    pub(crate) claim_tx_id: Option<HistoryTxId>,
+    pub(crate) mrh_tx_id: Option<HistoryTxId>,
+    pub(crate) mrh_amount_sat: Option<u64>, // TODO: Add mrh_amount recovery
```

Comment:
This is currently handled by calling the persister's `try_handle_receive_swap_update()` with the mrh_amount_set set. Do we need a different way to handle it?
https://github.com/breez/breez-sdk-liquid/blob/5f1fda010009c309d46f14e6b49e7b1ca9325e66/lib/core/src/persist/receive.rs#L263-L264